File: linked_lists/add_two_nums.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  carry = 0
  dummy = node = ListNode.new(0)

  while l1 || l2 || !count.zero?
    v1 = v2 = 0

    v1, l1 = l1.val, l1.next if l1
    v2, l2 = l2.val, l2.next if l2

    carry, val = (v1 + v2 + carry).divmod(10)
    node.next = ListNode.new(val)
    node = node.next
  end

  dummy.next
end

File: linked_lists/add_two_nums.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        dummy = curr = ListNode(None)

        while l1 or l2 or carry != 0:
            v1, v2 = 0, 0

            if l1: v1, l1 = l1.val, l1.next
            if l2: v2, l2 = l2.val, l2.next

            carry, val = divmod(v1 + v2 + carry, 10)
            curr.next = ListNode(val)
            curr = curr.next

        return dummy.next


# 2. Add Two Numbers
# https://leetcode.com/problems/add-two-numbers/description/
